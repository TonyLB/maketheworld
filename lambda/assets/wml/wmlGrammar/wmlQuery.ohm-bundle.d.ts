// AUTOGENERATED FILE
// This file was generated from wmlQuery.ohm by `ohm generateBundles`.

import {
  ActionDict,
  Grammar,
  IterationNode,
  Node,
  NonterminalNode,
  Semantics,
  TerminalNode
} from 'ohm-js';

export interface WMLQuerySelectorActionDict<T> extends ActionDict<T> {
  WMLQueryContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  wmlLegalTag?: (this: NonterminalNode, arg0: TerminalNode) => T;
  stringText?: (this: NonterminalNode, arg0: IterationNode) => T;
  legalProperties?: (this: NonterminalNode, arg0: TerminalNode) => T;
  propertyEqualityFilter?: (this: NonterminalNode, arg0: IterationNode, arg1: NonterminalNode, arg2: IterationNode, arg3: TerminalNode, arg4: NonterminalNode, arg5: TerminalNode) => T;
  propertyFilter?: (this: NonterminalNode, arg0: TerminalNode, arg1: NonterminalNode, arg2: TerminalNode) => T;
  firstFilter?: (this: NonterminalNode, arg0: TerminalNode) => T;
  nthChildFilter?: (this: NonterminalNode, arg0: TerminalNode, arg1: IterationNode, arg2: TerminalNode) => T;
  matchComponent?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  matchPredicate?: (this: NonterminalNode, arg0: IterationNode) => T;
  MatchFlatItem?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  MatchFlatList?: (this: NonterminalNode, arg0: IterationNode) => T;
  MatchGroup?: (this: NonterminalNode, arg0: TerminalNode, arg1: NonterminalNode, arg2: TerminalNode) => T;
  MatchOr?: (this: NonterminalNode, arg0: NonterminalNode, arg1: IterationNode) => T;
  MatchOrClause?: (this: NonterminalNode, arg0: TerminalNode, arg1: NonterminalNode) => T;
  MatchAncestry?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  whitespace_verticalTab?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace_formFeed?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace_noBreakSpace?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace_byteOrderMark?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace?: (this: NonterminalNode, arg0: NonterminalNode | TerminalNode) => T;
}

export interface WMLQuerySelectorSemantics extends Semantics {
  addOperation<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
  extendOperation<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
  addAttribute<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
  extendAttribute<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
}

export interface WMLQuerySelectorGrammar extends Grammar {
  createSemantics(): WMLQuerySelectorSemantics;
  extendSemantics(superSemantics: WMLQuerySelectorSemantics): WMLQuerySelectorSemantics;
}

declare const grammar: WMLQuerySelectorGrammar;
export default grammar;

