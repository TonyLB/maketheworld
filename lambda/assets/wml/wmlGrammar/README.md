The wml.ohm-bundle files are generated:  Only the wml.ohm and wmlQuery.ohm files are original.

To regenerate (e.g. after changing the wml.ohm file), navigate to the assets/wml directory, and
execute this command:  npx ohm generateBundles 'wmlGrammar/*.ohm'

... and then rename the 'ohm-bundle.js' files to 'ohm-bundle.cjs', since the --esm option for
CLI is not released yet, so we can only create CommonJS files, which Lambda needs to be flagged
for.