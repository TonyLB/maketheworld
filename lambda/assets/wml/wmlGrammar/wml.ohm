WorldMarkupLangage {

    //
    // This section defines the tags that make up the WML language
    //

    WMLFileContents = AssetExpression
            | StoryExpression
            | CharacterExpression

    CharacterExpression = TagExpression<"Character", "key", ("player" | "fileName" | "zone" | "subFolder"), none, none, CharacterLegalContents>
    CharacterLegalContents = ImageExpression
            | LiteralNameExpression
            | PronounsExpression
            | LiteralValueTag<"OneCoolThing">
            | LiteralValueTag<"Outfit">
            | LiteralValueTag<"FirstImpression">

    PronounsExpression = TagExpression<"Pronouns", none, ("subject" | "object" | "possessive" | "adjective" | "reflexive"), none, none, none>
            | TagSelfClosing<"Pronouns", none, ("subject" | "object" | "possessive" | "adjective" | "reflexive"), none, none>

    StoryExpression = TagExpression<"Story", "key", ("fileName" | "zone" | "subFolder" | "player"), none, "instance", AssetLegalContents>

    AssetExpression = TagExpression<"Asset", "key", ("fileName" | "zone" | "subFolder" | "player"), none, none, AssetLegalContents>
    AssetLegalContents = ImportExpression
            | VariableExpression
            | ComputedExpression
            | ActionExpression
            | LayerExpression
            | ConditionExpression
            | MapExpression
            | RoomExpression
            | ExitExpression
            | FeatureExpression
            | LiteralNameExpression
            | string

    MapExpression = TagExpression<"Map", "key", none, none, none, MapLegalContents>
    MapLegalContents = MapRoomExpression
            | NameExpression
            | ImageExpression

    ImageExpression = TagExpression<"Image", "key", "fileURL", none, none, none>
            | TagSelfClosing<"Image", "key", "fileURL", none, none>

    ImportExpression = TagExpression<"Import", "from", none, none, none, ImportLegalContents>
    ImportLegalContents = UseExpression

    UseExpression = TagExpression<"Use", ("key" | "as"), none, none, none, none>
            | TagSelfClosing<"Use", ("key" | "as"), none, none, none>

    DependencyExpression = TagExpression<"Depend", "on", none, none, none, none>
            | TagSelfClosing<"Depend", "on", none, none, none>

    VariableExpression = TagExpression<"Variable", "key", none, "default", none, none>
            | TagSelfClosing<"Variable", "key", none, "default", none>

    ComputedExpression = TagExpression<"Computed", "key", none, "src", none, DependencyExpression>
            | TagSelfClosing<"Computed", "key", none, "src", none>

    ActionExpression = TagExpression<"Action", "key", none, "src", none, none>
            | TagSelfClosing<"Action", "key", none, "src", none>

    ConditionExpression = TagExpression<"Condition", none, none, "if", none, ConditionLegalContents>
    ConditionLegalContents = RoomExpression
            | MapExpression
            | ConditionExpression
            | TextContents
            | LayerExpression
            | ExitExpression
            | FeatureExpression
            | DependencyExpression

    LayerExpression = TagExpression<"Layer", "key", none, none, none, LayerLegalContents>
    LayerLegalContents = RoomExpression
            | ConditionExpression
            | ExitExpression

    ExitExpression = TagExpression<"Exit", ("key" | "to" | "from"), none, none, none, string>
            | TagSelfClosing<"Exit", ("key" | "to" | "from"), none, none, none>

    LinkExpression = TagExpression<"Link", ("key" | "to"), none, none, none, string>

    MapRoomExpression = TagExpression<"Room", "key", ("display" | "x" | "y"), none, "global", RoomContents>
            | TagSelfClosing<"Room", "key", ("display" | "x" | "y"), none, "global">
    RoomExpression = TagExpression<"Room", "key", "display", none, "global", RoomContents>
            | TagSelfClosing<"Room", "key", "display", none, "global">
    RoomContents = TextContents
            | NameExpression
            | ExitExpression
            | FeatureExpression
            | LinkExpression

    FeatureExpression = TagExpression<"Feature", "key", "display", none, "global", FeatureContents>
            | TagSelfClosing<"Feature", "key", "display", none, "global">
    FeatureContents = TextContents
            | NameExpression
            | LinkExpression

    LiteralNameExpression = LiteralValueTag<"Name">
    NameExpression = ValueTag<"Name">

    LiteralValueTag<label> = TagExpression<label, none, none, none, none, string>

    ValueTag<label> = TagExpression<label, none, none, none, none, TextContents>
    TextContents = string | EmbeddedJSExpression

    //
    // The following define the ways in which tags can be structured
    //

    TagExpression<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs, LegalContents> = TagOpen<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> LegalContents* TagClose<tag>
    string = (stringText | spaceCompressor)+
    stringText = (~("<" | "{" | space) any | "\\<" | "\\{")+
    legalKey = (letter | digit | "-" | "_")*
    spaceCompressor = (space)+
    TagSelfClosing<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> = "<" tag TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs>* "/>"
    TagOpen<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> = "<" tag TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs>* ">"
    TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> = TagArgumentKey<tagKeyArgs> | TagLiteralArgument<tagLiteralArgs> | TagProgramArgument<tagProgramArgs> | tagBooleanArgument<tagBooleanArgs>
    TagLiteralArgument<tagLiteralArgs> = tagArgumentQuoted<tagLiteralArgs>
    TagProgramArgument<tagProgramArgs> = tagArgumentQuoted<tagProgramArgs> | TagArgumentBracketed<tagProgramArgs>
    tagBooleanArgument<tagBooleanArgs> = tagBooleanArgs &(space | ">" | "/>")
    tagArgumentQuoted<tagLegalArgs> = tagLegalArgs "=\"" tagArgValueQuoted
    TagArgumentKey<tagLegalArgs> = #(tagLegalArgs "=(") legalKey ")"
    TagArgumentBracketed<tagLegalArgs> = #(tagLegalArgs "={") JSExpression* "}"
    tagArgValueQuoted = ("\\\"" | ~"\"" any)* "\""
    TagClose<tag> = "</" tag ">"
    none = ~">" ">"

    //
    // EmbeddedJSExpression parses from the starting open bracket through to the
    //   *MATCHING* close bracket, using a subset of JS syntax that concentrates
    //   on all the ways that JS could let you mess up that matching (quotes,
    //   nesting, and template literals figuring high on the list)
    //
    // While we could, technically, validate the entire ES5 JS Syntax (after all,
    // Ohm publishes their schema), it wouldn't gain us any value here, and it
    // has a quite noticeable resource cost:  Compiling that schema and executing
    // it slows the perception Lambda function down to a crawl.  If a need for
    // such pre-processing emerges, look into how to optimize compiling and executing
    // very large Ohm schemata.
    //
    EmbeddedJSExpression = "{" JSExpression* "}"
    JSExpression = EmbeddedJSExpression
                    | JSDoubleQuotedString
                    | JSSingleQuotedString
                    | JSTemplateString
                    | JSText
    JSDoubleQuotedString = "\"" JSNoDoubleQuote* "\""
    JSNoDoubleQuote = "\\\"" | (~"\"" any)
    JSSingleQuotedString = "\'" JSNoSingleQuote* "\'"
    JSNoSingleQuote = "\\\'" | (~"\'" any)
    JSTemplateString = "`" (JSNoBackQuote | JSTemplateEvaluation )* "`"
    JSTemplateEvaluation = "${" JSExpression "}"
    JSNoBackQuote = "\\`" | (~("`" | "${") any)
    JSText = ("\\{" | "\\\"" | "\\'" | "\\`" | "\\}" | (~("{" | "\"" | "\'" | "`" | "}") any))+

    // Override Ohm's built-in definition of space.  Cribbed from es5.ohm
    space := whitespace | lineTerminator | comment

	sourceCharacter = any
    whitespace = "\t"
                | "\x0B"    -- verticalTab
                | "\x0C"    -- formFeed
                | " "
                | "\u00A0"  -- noBreakSpace
                | "\uFEFF"  -- byteOrderMark
                | unicodeSpaceSeparator

	unicodeSpaceSeparator = "\u2000".."\u200B" | "\u3000"

    lineTerminator = "\n" | "\r" | "\u2028" | "\u2029"
    lineTerminatorSequence = "\n" | "\r" ~"\n" | "\u2028" | "\u2029" | "\r\n"

    comment = multiLineComment | singleLineComment

    multiLineComment = "/*" (~"*/" sourceCharacter)* "*/"
    singleLineComment = "//" (~lineTerminator sourceCharacter)*

}
