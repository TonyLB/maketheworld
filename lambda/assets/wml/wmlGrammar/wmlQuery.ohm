WMLQuerySelector {
    WMLQueryContents = TagAncestry

    WMLLegalTag = "Asset" |
        "Character" |
        "Name" |
        "Pronouns" |
        "OneCoolThing" |
        "Outfit" |
        "FirstImpression" |
        "Condition" |
        "Layer" |
        "Exit" |
        "Room"

    TagAncestry = WMLLegalTag+

}