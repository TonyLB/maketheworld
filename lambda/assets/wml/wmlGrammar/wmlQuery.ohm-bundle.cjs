'use strict';const ohm=require('ohm-js');module.exports=ohm.makeRecipe(["grammar",{"source":"WMLQuerySelector {\r\n    WMLQueryContents = TagAncestry\r\n\r\n    WMLLegalTag = \"Asset\" |\r\n        \"Character\" |\r\n        \"Name\" |\r\n        \"Pronouns\" |\r\n        \"OneCoolThing\" |\r\n        \"Outfit\" |\r\n        \"FirstImpression\" |\r\n        \"Condition\" |\r\n        \"Layer\" |\r\n        \"Exit\" |\r\n        \"Room\"\r\n\r\n    TagAncestry = WMLLegalTag+\r\n\r\n}"},"WMLQuerySelector",null,"WMLQueryContents",{"WMLQueryContents":["define",{"sourceInterval":[24,54]},null,[],["app",{"sourceInterval":[43,54]},"TagAncestry",[]]],"WMLLegalTag":["define",{"sourceInterval":[62,299]},null,[],["alt",{"sourceInterval":[76,299]},["terminal",{"sourceInterval":[76,83]},"Asset"],["terminal",{"sourceInterval":[95,106]},"Character"],["terminal",{"sourceInterval":[118,124]},"Name"],["terminal",{"sourceInterval":[136,146]},"Pronouns"],["terminal",{"sourceInterval":[158,172]},"OneCoolThing"],["terminal",{"sourceInterval":[184,192]},"Outfit"],["terminal",{"sourceInterval":[204,221]},"FirstImpression"],["terminal",{"sourceInterval":[233,244]},"Condition"],["terminal",{"sourceInterval":[256,263]},"Layer"],["terminal",{"sourceInterval":[275,281]},"Exit"],["terminal",{"sourceInterval":[293,299]},"Room"]]],"TagAncestry":["define",{"sourceInterval":[307,333]},null,[],["plus",{"sourceInterval":[321,333]},["app",{"sourceInterval":[321,332]},"WMLLegalTag",[]]]]}]);