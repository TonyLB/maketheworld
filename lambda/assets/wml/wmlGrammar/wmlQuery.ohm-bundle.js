import ohm from 'ohm-js';const result=ohm.makeRecipe(["grammar",{"source":"WMLQuerySelector {\r\n    WMLQueryContents = MatchAncestry\r\n\r\n    wmlLegalTag = \"Asset\" |\r\n        \"Character\" |\r\n        \"Name\" |\r\n        \"Pronouns\" |\r\n        \"OneCoolThing\" |\r\n        \"Outfit\" |\r\n        \"FirstImpression\" |\r\n        \"Condition\" |\r\n        \"Layer\" |\r\n        \"Exit\" |\r\n        \"Description\" |\r\n        \"Room\" |\r\n        \"Map\"\r\n\r\n    stringText = (~(\"<\" | \"{\" | \"\\\"\") any | \"\\\\<\" | \"\\\\{\" | \"\\\\\\\"\")+\r\n\r\n    legalProperties = \"key\" |\r\n        \"to\" |\r\n        \"from\"\r\n\r\n    propertyEqualityFilter = whitespace* legalProperties whitespace* \"=\\\"\" stringText \"\\\"\"\r\n\r\n    propertyFilter = \"[\" propertyEqualityFilter \"]\"\r\n\r\n    firstFilter = \":first\"\r\n\r\n    nthChildFilter = \":nthChild(\" digit* \")\"\r\n\r\n    matchComponent = wmlLegalTag |\r\n        propertyFilter |\r\n        firstFilter |\r\n        nthChildFilter\r\n\r\n    matchPredicate = matchComponent+\r\n\r\n    MatchAncestry = matchPredicate+\r\n\r\n    whitespace = \"\\t\"\r\n        | \"\\x0B\"    -- verticalTab\r\n        | \"\\x0C\"    -- formFeed\r\n        | \" \"\r\n        | \"\\u00A0\"  -- noBreakSpace\r\n        | \"\\uFEFF\"  -- byteOrderMark\r\n\r\n}"},"WMLQuerySelector",null,"WMLQueryContents",{"WMLQueryContents":["define",{"sourceInterval":[24,56]},null,[],["app",{"sourceInterval":[43,56]},"MatchAncestry",[]]],"wmlLegalTag":["define",{"sourceInterval":[64,343]},null,[],["alt",{"sourceInterval":[78,343]},["terminal",{"sourceInterval":[78,85]},"Asset"],["terminal",{"sourceInterval":[97,108]},"Character"],["terminal",{"sourceInterval":[120,126]},"Name"],["terminal",{"sourceInterval":[138,148]},"Pronouns"],["terminal",{"sourceInterval":[160,174]},"OneCoolThing"],["terminal",{"sourceInterval":[186,194]},"Outfit"],["terminal",{"sourceInterval":[206,223]},"FirstImpression"],["terminal",{"sourceInterval":[235,246]},"Condition"],["terminal",{"sourceInterval":[258,265]},"Layer"],["terminal",{"sourceInterval":[277,283]},"Exit"],["terminal",{"sourceInterval":[295,308]},"Description"],["terminal",{"sourceInterval":[320,326]},"Room"],["terminal",{"sourceInterval":[338,343]},"Map"]]],"stringText":["define",{"sourceInterval":[351,415]},null,[],["plus",{"sourceInterval":[364,415]},["alt",{"sourceInterval":[365,413]},["seq",{"sourceInterval":[365,388]},["not",{"sourceInterval":[365,384]},["alt",{"sourceInterval":[367,383]},["terminal",{"sourceInterval":[367,370]},"<"],["terminal",{"sourceInterval":[373,376]},"{"],["terminal",{"sourceInterval":[379,383]},"\""]]],["app",{"sourceInterval":[385,388]},"any",[]]],["terminal",{"sourceInterval":[391,396]},"\\<"],["terminal",{"sourceInterval":[399,404]},"\\{"],["terminal",{"sourceInterval":[407,413]},"\\\""]]]],"legalProperties":["define",{"sourceInterval":[423,480]},null,[],["alt",{"sourceInterval":[441,480]},["terminal",{"sourceInterval":[441,446]},"key"],["terminal",{"sourceInterval":[458,462]},"to"],["terminal",{"sourceInterval":[474,480]},"from"]]],"propertyEqualityFilter":["define",{"sourceInterval":[488,574]},null,[],["seq",{"sourceInterval":[513,574]},["star",{"sourceInterval":[513,524]},["app",{"sourceInterval":[513,523]},"whitespace",[]]],["app",{"sourceInterval":[525,540]},"legalProperties",[]],["star",{"sourceInterval":[541,552]},["app",{"sourceInterval":[541,551]},"whitespace",[]]],["terminal",{"sourceInterval":[553,558]},"=\""],["app",{"sourceInterval":[559,569]},"stringText",[]],["terminal",{"sourceInterval":[570,574]},"\""]]],"propertyFilter":["define",{"sourceInterval":[582,629]},null,[],["seq",{"sourceInterval":[599,629]},["terminal",{"sourceInterval":[599,602]},"["],["app",{"sourceInterval":[603,625]},"propertyEqualityFilter",[]],["terminal",{"sourceInterval":[626,629]},"]"]]],"firstFilter":["define",{"sourceInterval":[637,659]},null,[],["terminal",{"sourceInterval":[651,659]},":first"]],"nthChildFilter":["define",{"sourceInterval":[667,707]},null,[],["seq",{"sourceInterval":[684,707]},["terminal",{"sourceInterval":[684,696]},":nthChild("],["star",{"sourceInterval":[697,703]},["app",{"sourceInterval":[697,702]},"digit",[]]],["terminal",{"sourceInterval":[704,707]},")"]]],"matchComponent":["define",{"sourceInterval":[715,818]},null,[],["alt",{"sourceInterval":[732,818]},["app",{"sourceInterval":[732,743]},"wmlLegalTag",[]],["app",{"sourceInterval":[755,769]},"propertyFilter",[]],["app",{"sourceInterval":[781,792]},"firstFilter",[]],["app",{"sourceInterval":[804,818]},"nthChildFilter",[]]]],"matchPredicate":["define",{"sourceInterval":[826,858]},null,[],["plus",{"sourceInterval":[843,858]},["app",{"sourceInterval":[843,857]},"matchComponent",[]]]],"MatchAncestry":["define",{"sourceInterval":[866,897]},null,[],["plus",{"sourceInterval":[882,897]},["app",{"sourceInterval":[882,896]},"matchPredicate",[]]]],"whitespace_verticalTab":["define",{"sourceInterval":[934,958]},null,[],["terminal",{"sourceInterval":[934,940]},"\u000b"]],"whitespace_formFeed":["define",{"sourceInterval":[970,991]},null,[],["terminal",{"sourceInterval":[970,976]},"\f"]],"whitespace_noBreakSpace":["define",{"sourceInterval":[1018,1043]},null,[],["terminal",{"sourceInterval":[1018,1026]}," "]],"whitespace_byteOrderMark":["define",{"sourceInterval":[1055,1081]},null,[],["terminal",{"sourceInterval":[1055,1063]},"﻿"]],"whitespace":["define",{"sourceInterval":[905,1081]},null,[],["alt",{"sourceInterval":[918,1081]},["terminal",{"sourceInterval":[918,922]},"\t"],["app",{"sourceInterval":[934,940]},"whitespace_verticalTab",[]],["app",{"sourceInterval":[970,976]},"whitespace_formFeed",[]],["terminal",{"sourceInterval":[1003,1006]}," "],["app",{"sourceInterval":[1018,1026]},"whitespace_noBreakSpace",[]],["app",{"sourceInterval":[1055,1063]},"whitespace_byteOrderMark",[]]]]}]);export default result;