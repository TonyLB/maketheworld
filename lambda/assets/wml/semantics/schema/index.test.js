import wmlGrammar from '../../wmlGrammar/wml.ohm-bundle.js'
import { wmlSemantics } from '../../index.js'

describe('WML semantic schema', () => {
    it('should parse elements properly', () => {
        const match = wmlGrammar.match(`
            <Asset key=(Test) fileName="test">
                <Import from=(BASE)>
                    <Use key=(basePower) as=(power) />
                    <Use key=(overview) />
                </Import>
                <Room key=(ABC)>
                    <Name>Vortex</Name>
                    Vortex
                    <Link key=(switchToggle) to=(toggleOpen)>(toggle)</Link>
                    <Exit from=(DEF)>vortex</Exit>
                </Room>
                <Condition if={open}>
                    <Depend on=(open) />
                    <Room key=(ABC)>
                        <Exit to=(DEF)>welcome</Exit>
                    </Room>
                </Condition>
                <Room key=(DEF)>
                    <Name>Welcome</Name>
                </Room>
                <Variable key=(open) default={false} />
                <Action key=(toggleOpen) src={open = !open} />
                <Computed key=(closed) src={!open}>
                    <Depend on=(open) />
                </Computed>
            </Asset>
        `)
        const schema = wmlSemantics(match).schema()
        expect(schema).toEqual({
            key: 'Test',
            tag: 'Asset',
            fileName: 'test',
            props: {},
            contents: [{
                tag: 'Import',
                from: 'BASE',
                mapping: {
                    power: 'basePower',
                    overview: 'overview'
                },
                contents: [],
                props: {}
            },
            {
                key: 'ABC',
                tag: 'Room',
                name: 'Vortex',
                global: false,
                render: [
                    'Vortex ',
                    {
                        tag: 'Link',
                        key: 'switchToggle',
                        text: '(toggle)',
                        to: 'toggleOpen',
                        props: {},
                        contents: []
                    }
                ],
                props: {},
                contents: [{
                    tag: 'Exit',
                    from: 'DEF',
                    name: 'vortex',
                    props: {},
                    contents: []
                }]
            },
            {
                tag: 'Condition',
                if: 'open',
                dependencies: ['open'],
                props: {},
                contents: [{
                    key: 'ABC',
                    tag: 'Room',
                    global: false,
                    render: [],
                    props: {},
                    contents: [{
                        tag: 'Exit',
                        to: 'DEF',
                        name: 'welcome',
                        props: {},
                        contents: []
                    }]
                }]
            },
            {
                key: 'DEF',
                tag: 'Room',
                name: 'Welcome',
                global: false,
                props: {},
                render: [],
                contents: []
            },
            {
                key: 'open',
                tag: 'Variable',
                default: 'false',
                props: {}
            },
            {
                key: 'toggleOpen',
                tag: 'Action',
                src: 'open = !open',
                props: {}
            },
            {
                key: 'closed',
                tag: 'Computed',
                dependencies: ['open'],
                src: '!open',
                props: {}
            }]
        })
    })
})