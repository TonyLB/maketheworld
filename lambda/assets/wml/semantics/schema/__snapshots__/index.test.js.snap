// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WML semantic schema should derive spaceBefore and spaceAfter properties from whitespace in description 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "stone",
      "props": Object {},
      "render": Array [],
      "tag": "Feature",
    },
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "clockTower",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "A clock-tower of weathered grey-",
        },
        Object {
          "contents": Array [],
          "key": "towerStone",
          "props": Object {},
          "spaceAfter": true,
          "spaceBefore": false,
          "tag": "Link",
          "text": "stone",
          "to": "stone",
        },
        Object {
          "spaceAfter": false,
          "spaceBefore": true,
          "tag": "String",
          "value": "looms over the area.",
        },
      ],
      "tag": "Feature",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "clockTower",
          "props": Object {},
          "render": Array [],
          "tag": "Feature",
        },
      ],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": true,
          "spaceBefore": false,
          "tag": "String",
          "value": "A spinning tumble of wreckage, surrounding a",
        },
        Object {
          "contents": Array [],
          "key": "vortexClockTower",
          "props": Object {},
          "spaceAfter": false,
          "spaceBefore": true,
          "tag": "Link",
          "text": "clock tower",
          "to": "clockTower",
        },
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": ".",
        },
      ],
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse a character element 1`] = `
Object {
  "Name": "Tess",
  "Pronouns": Object {
    "adjective": "hers",
    "object": "her",
    "possessive": "her",
    "reflexive": "herself",
    "subject": "she",
  },
  "contents": Array [],
  "fileName": "Tess",
  "fileURL": "testIcon.png",
  "key": "Tess",
  "player": "testy",
  "props": Object {},
  "tag": "Character",
  "zone": "Library",
}
`;

exports[`WML semantic schema should parse a story element 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "name": "Vortex",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "Vortex",
        },
      ],
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse elements properly 1`] = `
Object {
  "contents": Array [
    Object {
      "contents": Array [],
      "from": "BASE",
      "mapping": Object {
        "overview": Object {
          "key": "overview",
          "type": "Room",
        },
        "power": Object {
          "key": "basePower",
          "type": "Variable",
        },
      },
      "props": Object {},
      "tag": "Import",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "from": "DEF",
          "key": undefined,
          "name": "vortex",
          "props": Object {},
          "tag": "Exit",
          "to": undefined,
        },
      ],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "name": "Vortex",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": true,
          "spaceBefore": false,
          "tag": "String",
          "value": "Vortex",
        },
        Object {
          "contents": Array [],
          "key": "switchToggle",
          "props": Object {},
          "spaceAfter": false,
          "spaceBefore": true,
          "tag": "Link",
          "text": "(toggle)",
          "to": "toggleOpen",
        },
      ],
      "tag": "Room",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [
            Object {
              "contents": Array [],
              "from": undefined,
              "key": undefined,
              "name": "welcome",
              "props": Object {},
              "tag": "Exit",
              "to": "DEF",
            },
          ],
          "display": undefined,
          "global": false,
          "key": "ABC",
          "props": Object {},
          "render": Array [],
          "tag": "Room",
        },
      ],
      "dependencies": Array [
        "open",
      ],
      "if": "open",
      "props": Object {},
      "tag": "Condition",
    },
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "DEF",
      "name": "Welcome",
      "props": Object {},
      "render": Array [],
      "tag": "Room",
    },
    Object {
      "default": "false",
      "key": "open",
      "props": Object {},
      "tag": "Variable",
    },
    Object {
      "key": "toggleOpen",
      "props": Object {},
      "src": "open = !open",
      "tag": "Action",
    },
    Object {
      "dependencies": Array [
        "open",
      ],
      "key": "closed",
      "props": Object {},
      "src": "!open",
      "tag": "Computed",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse feature elements 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "clockTower",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "A clock-tower of weathered grey stone looms over the area.",
        },
      ],
      "tag": "Feature",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "clockTower",
          "props": Object {},
          "render": Array [],
          "tag": "Feature",
        },
      ],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "name": "Vortex",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "Vortex",
        },
      ],
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse map elements 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "fileURL": "https://test.com/imageTest.png",
          "key": "ImageTest",
          "props": Object {},
          "tag": "Image",
        },
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "ABC",
          "props": Object {},
          "render": Array [],
          "tag": "Room",
        },
      ],
      "key": "TestMap",
      "name": "Test Map",
      "props": Object {},
      "rooms": Object {
        "ABC": Object {
          "x": 200,
          "y": 150,
        },
      },
      "tag": "Map",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;
